crates/codegen/runtime/generator/src/kinds.rs

```diff
@@ -0,0 +1,98 @@
+use std::collections::BTreeSet;
+
+use codegen_language_definition::model::{self, Identifier, Item};
+use serde::Serialize;
+
+#[derive(Default, Serialize)]
+pub struct KindsModel {
+    /// Defines the `NonterminalKind` enum variants.
+    nonterminal_kinds: BTreeSet<Identifier>,
+    /// Defines the `TerminalKind` enum variants.
+    terminal_kinds: BTreeSet<Identifier>,
+    /// Defines `TerminalKind::is_trivia` method.
+    trivia_scanner_names: BTreeSet<Identifier>,
+    /// Defines `EdgeLabel` enum variants.
+    labels: BTreeSet<Identifier>,
+    // Defines the `LexicalContext(Type)` enum and type-level variants.
+    lexical_contexts: BTreeSet<Identifier>,
+}
+
+impl KindsModel {
+    pub fn create(language: &model::Language) -> Self {
+        let terminal_kinds = language
```

Comment:
I think there are a few places in the codebase that can benefit from such API to collect these kinds, not just the runtime/parser. For example:

- https://github.com/xanewok/slang/blob/2d8b7ddac1d457c43e908b56f356d056d0730507/crates/codegen/language/definition/src/compiler/analysis/reachability.rs#L56
- https://github.com/xanewok/slang/blob/2d8b7ddac1d457c43e908b56f356d056d0730507/crates/codegen/runtime/generator/src/ast.rs#L64
- https://github.com/xanewok/slang/blob/2d8b7ddac1d457c43e908b56f356d056d0730507/crates/codegen/runtime/generator/src/ast.rs#L67

Thoughts about moving this collector to a helper methods in the original language crate instead of here?

- `Language.collect_terminals() -> Vec<&Identifier>`
- `Language.collect_non_terminals() -> Vec<&Identifier>`
- `Language.collect_trivia() -> Vec<&Identifier>`
- etc...

I'm happy to refactor validation/AST code later to use such APIs, and remove the duplicate code.

Reply:
Yes, it might be beneficial to provide some helper API to conveniently iterate over the items in the language but I'd prefer if we can create a composable and readable API, e.g. `language.items().filter(Item::is_terminal)` rather than hand-roll multiple variants that differ very slightly (unless it's very common, think 4-5+ call sites).

Let's expand the use of `Language` first and revisit this once we will have more concrete use cases.

Reply:
Done once in 44b4ac5fc62e9a28aa194e1680543b22fa903c80 for now, now that we made the change from https://github.com/NomicFoundation/slang/pull/991#discussion_r1624501253

Reply:
The only exception to that is `OperatorDefinition`, which contributes non-terminals, but is not an item, so it won't be possible to collect it using `language.items().filter(...)`.
But otherwise, sounds good to me. Not a blocker of course.

Reply:
The only exception to that is `PrecedenceExpression`, which contributes non-terminals, but is not an item, so it won't be possible to collect it using `language.items().filter(...)`. So we need a specific API for collecting them to handle such cases.

But otherwise, sounds good to me. Not a blocker of course.
I'm refactoring the AST model you updated in #1000, so I can rebase on your PR, and look into this refactoring afterwards, in order not to block you.

Reply:
> So we need a specific API for collecting them to handle such cases.

That's fine, I was simply commenting on the example API that collects the names of e.g. trivia, terminals in which case it's probably better to use a more composable approach :thumbsup:

> I'm refactoring the AST model you updated in https://github.com/NomicFoundation/slang/pull/1000

Ah, okay then; I'll just drop the helper changes from this PR not to mess with your flow (it's already a drive-by thing 😅 )